src/fusets/peakvalley.py

```diff
@@ -0,0 +1,168 @@
+from datetime import datetime
+from typing import Iterable, Sequence, Tuple
+
+import numpy as np
+import xarray
+from scipy.signal import find_peaks
+from xarray import DataArray
+
+from fusets._xarray_utils import _extract_dates, _time_dimension
+
+
+def peakvalley(
+    array: DataArray,
+    drop_thr: float = 0.15,
+    rec_r: float = 1.0,
+    slope_thr: float = -0.007,
+) -> DataArray:
+
+    dates = np.array(_extract_dates(array))
+    time_dimension = _time_dimension(array, None)
+
+    def callback(timeseries):
+        out, _ = peakvalley_f(dates, timeseries, drop_thr, rec_r, slope_thr)
+        return out
+
+    result = xarray.apply_ufunc(
+        callback,
+        array,
+        input_core_dims=[[time_dimension]],
+        output_core_dims=[[time_dimension]],
+        vectorize=True,
+    )
+
+    result = result.rename("peak_valley_mask")
+    return result
+
+
+def peakvalley_f(
+    x: Sequence[datetime],
+    y: np.ndarray,
+    drop_thr: float = 0.15,
+    rec_r: float = 1.0,
+    slope_thr: float = -0.007,
+) -> np.ndarray:
+    """
+    Algorithm for finding peak-valley patterns in the provided array.
+
+    @param x: array of timestamps
+    @param y: array of input feature values
+    @param drop_thr: threshold value for the amplitude of the drop in the input feature
+    @param rec_r: threshold value for the amplitude of the recovery, relative to the `drop_delta`
+    @param slope_thr: threshold value for the slope where the peak should start
+    @return: array with different values
+                *  1: peak
+                * -1: valley
+                *  0: between peak and valley
+                * np.nan: values outside of found peak-valley patterns
+    """
+
+    drop_thr, rec_thr = drop_thr, drop_thr * rec_r
+    result = np.full_like(y, np.nan)
+
+    # find peaks and valleys in trend
+    nan_mask = np.isnan(y)
+    feature = y[~nan_mask]
+    timestamps = x[~nan_mask]
+
+    pk_ids = find_peaks(feature)[0]
+    vl_ids = find_peaks(-feature)[0]
+    if len(pk_ids) == 0 or len(vl_ids) == 0:
```

Comment:
Can it happen that we have a peak and not a valley or vice-versa?

Reply:
No, this has already been handled in lines https://github.com/Open-EO/FuseTS/blob/feat/peak-valley-algo/src/fusets/peakvalley.py#L73-L79, so that each peak has it's own valley and they both come in pairs